src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,296 @@
+import {
+	DeviceLog,
+	DeviceLogsBackend,
+	LogContext,
+	LogWriteContext,
+	Subscription,
+} from '../struct';
+import { captureException } from '../../../platform/errors';
+
+const { BadRequestError } = errors;
+
+const MIN_BACKOFF = 100;
+const MAX_BACKOFF = 10 * 1000;
+const VERSION = 1;
+
+function createTimestampFromDate(date = new Date()) {
+	const timestamp = new Timestamp();
+	timestamp.fromDate(date);
+	return timestamp;
+}
+
+function createTimestampFromMilliseconds(milliseconds: number) {
+	const MS_TO_NANOS = 1000000;
+	const seconds = Math.floor(milliseconds / 1000);
+	const nanos = (milliseconds - seconds * 1000) * MS_TO_NANOS;
+
+	const timestamp = new Timestamp();
+	timestamp.setSeconds(seconds);
+	timestamp.setNanos(nanos);
+	return timestamp;
+}
+
+export class LokiBackend implements DeviceLogsBackend {
+	private subscriptions: EventEmitter;
+	private querier: QuerierClient;
+	private pusher: IGrpcClientAsync;
+	private ready: boolean;
+	private tailCalls: Map<string, grpc.ClientReadableStream<TailResponse>>;
+
+	constructor() {
+		this.subscriptions = new EventEmitter();
+		this.querier = new QuerierClient(
+			`${LOKI_HOST}:${LOKI_PORT}`,
+			createInsecureCredentials(),
+		);
+		this.pusher = promisifyClient(
+			new PusherClient(
+				`${LOKI_HOST}:${LOKI_PORT}`,
+				createInsecureCredentials(),
+			),
+		);
+		this.ready = false;
+		this.tailCalls = new Map();
+	}
+
+	public get available(): boolean {
+		return !this.ready;
+	}
+
+	/**
+	 *
+	 * Return $count of logs matching device_id in descending (BACKWARD) order.
+	 *
+	 * The logs are sorted by timestamp since Loki returns a distinct stream for each label combination.
+	 *
+	 * @param ctx
+	 * @param count
+	 */
+	public async history(ctx: LogContext, count: number): Promise<DeviceLog[]> {
+		const oneHourAgo = new Date(Date.now() - 10000 * 60);
+
+		const queryRequest = new QueryRequest();
+		queryRequest.setSelector(this.getDeviceQuery(ctx));
+		queryRequest.setLimit(Number.isFinite(count) ? count : 1000);
+		queryRequest.setStart(createTimestampFromDate(oneHourAgo));
+		queryRequest.setEnd(createTimestampFromDate());
+		queryRequest.setDirection(Direction.BACKWARD);
+
+		const streams: StreamAdapter[] = [];
+		const call = this.querier.query(
+			queryRequest,
+			createOrgIdMetadata(String(ctx.belongs_to__application)),
+		);
+		const responseStreams: StreamAdapter[] = await new Promise(
+			(resolve, reject) => {
+				call.on('data', (queryResponse: QueryResponse) => {
+					streams.push(...queryResponse.getStreamsList());
+				});
+				call.on('error', (error: Error & { details: string }) => {
+					captureException(error);
+					reject(
+						new BadRequestError(`Failed to query logs for device ${ctx.uuid}`),
+					);
+				});
+				call.on('end', () => {
+					resolve(streams);
+				});
+			},
+		);
+		return _.orderBy(
+			this.fromStreamsToDeviceLogs(responseStreams),
+			'timestamp',
+			'desc',
+		);
```

Comment:
Am I right in understand that when getting history we return the newest entry first and the oldest last? That seems strange to me - is that what the redis backend does?

Reply:
Your right the order is incorrect. I have updated it. The logic for fetching `history` will likely change before we switch to reading from Loki, the reason is that with redis it was only required to request all the logs stored, where with Loki we want to display the last N logs but its uncertain at what time period those logs occurred. So multiple requests scanning backwards might be necessary. I have set the `LokiBackend` history to just retrieve all logs from the last hour, but for devices that don't log very often it may return just a few or 0 logs.